megamek/src/megamek/client/ui/baseComponents/AbstractDialog.java

```diff
@@ -0,0 +1,212 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MegaMek. If not, see <http://www.gnu.org/licenses/>.
+ */
+package megamek.client.ui.baseComponents;
+
+import megamek.MegaMek;
+import megamek.common.util.EncodeControl;
+import megamek.client.ui.preferences.JWindowPreference;
+import megamek.client.ui.preferences.PreferencesNode;
+
+import javax.swing.*;
+import java.awt.*;
+import java.awt.event.*;
+import java.util.ResourceBundle;
+
+/**
+ * This is the base class for dialogs in MegaMek. This class handles setting the UI, managing the X
+ * button, managing the escape key, and saving the dialog preferences.
+ *
+ * Inheriting classes must call initialize() in their constructor and override createCenterPane()
+ *
+ * This is directly tied to MekHQ's AbstractMHQDialog, and any changes here MUST be verified there.
+ */
+public abstract class AbstractDialog extends JDialog implements WindowListener {
+    //region Variable Declarations
+    private JFrame frame;
+
+    protected static final String CLOSE_ACTION = "closeAction";
+
+    protected ResourceBundle resources;
+    //endregion Variable Declarations
+
+    //region Constructors
+    protected AbstractDialog(final JFrame frame, final String name, final String title) {
+        this(frame, false, name, title);
+    }
+
+    protected AbstractDialog(final JFrame frame, final boolean modal, final String name, final String title) {
+        this(frame, modal, ResourceBundle.getBundle("megamek.client.messages", new EncodeControl()), name, title);
+    }
+
+    protected AbstractDialog(final JFrame frame, final boolean modal, final ResourceBundle resources,
+                             final String name, final String title) {
+        super(frame, modal);
+        setTitle(resources.getString(title));
+        setName(name);
+        setFrame(frame);
+        this.resources = resources;
+    }
+    //endregion Constructors
+
+    //region Getters/Setters
+    public JFrame getFrame() {
+        return frame;
+    }
+
+    public void setFrame(final JFrame frame) {
+        this.frame = frame;
+    }
+    //endregion Getters/Setters
+
+    //region Initialization
+    /**
+     * Initializes the dialog's UI and preferences. Needs to be called by child classes for initial
+     * setup.
+     */
+    protected void initialize() {
+        setLayout(new BorderLayout());
+        add(createCenterPane(), BorderLayout.CENTER);
+        finalizeInitialization();
+    }
+
+    /**
+     * This is used to create the dialog's center pane
+     * @return the center pane of the dialog
+     */
+    protected abstract Container createCenterPane();
+
+    /**
+     * This MUST be called at the end of initialization to finalize it. This is the key method for
+     * this being the abstract basis for all other dialog
+     */
+    protected void finalizeInitialization() {

```

Comment:
Is there a way that we can force this method to be called, rather than having to rely on the developer to do it? Two years from now, someone's going to create an AbstractDialog, not know to call this, then spend a bunch of time wondering why their dialog doesn't work.

Reply:
I've added notes in the JavaDoc for both AbstractDialog and AbstractButtonDialog, and both call it to end initialize. I can't think of any other way to handle it other than expecting the dev to read.

Reply:
I think it's a reasonable ask devs to call initialize. Another way to help make this more clear to future devs, would be to add that information also in the documentation of createCenterPanel, as the dev using this dialog needs to override that method to make the code compile.

Reply:
This is normally included in initialize, just causes problems if someone overrides initialize.